File: AoC2019/Day01.fs
namespace AoC2019

open System
open System.IO

module Day01 = 

    let basePath : string = AppContext.BaseDirectory.Replace("CLI", "AoC2019") // check to make sure when getting called from cli .. could probably be nicer

    let input file = File.ReadLines(basePath + file) |> Seq.map(fun line -> line |> int )

    let GetFuel (mass : int) : int = Math.Floor((float mass) / 3.0 ) - 2.0 |> int

    let GetFuelRecursive(mass : int) : int = 
        let rec loop mass total =
            if mass > 6 then
                let fuelMass = GetFuel mass
                loop fuelMass total + fuelMass
            else total
        loop mass 0


    let SolvePart1 : string = input("data/day01.txt") |> Seq.map(fun i -> GetFuel(i) ) |> Seq.sum |> string

    let SolvePart2 : string = input("data/day01.txt") |> Seq.map(fun i -> GetFuelRecursive(i) ) |> Seq.sum |> string

File: AoC2019/Day01.cs
namespace AoC2019;

public class Day01(string file) : Solution(file, "\n")
{
	public override async Task<string> SolvePart1() => Input.Select(int.Parse).Select(GetFuelPerModule).Sum().ToString();

	public override async Task<string> SolvePart2() => Input.Select(int.Parse).Select(mass =>
	{
		var fuelMass = GetFuelPerModule(mass);
		var total = fuelMass;
		while (fuelMass > 0)
		{
			fuelMass = GetFuelPerModule(fuelMass);
			if (fuelMass > 0) total += fuelMass;
		}
		return total;
	}).Sum().ToString();

	public static int GetFuelPerModule(int mass) => mass / 3 - 2;

}